## LightGBM with Simple Feature Engineering
To Do:
- [x] Add 4 weeks of data
- [x] Add 8 weeks of data
- [x] Add feature importance\selection 4 weeks of data
- [x] Add feature importance\selection 8 weeks of data
- [x] Add 4 monhts of data to perfrom 70/30 split (best results so far)
- [] Add hyperparameter tuning (next nb)

### 4 weeks of data

In [1]:
import pandas as pd
from src.paths import TRANSFORMED_DATA_DIR

df = pd.read_parquet(TRANSFORMED_DATA_DIR / 'tabular_data.parquet')
df

rides_previous_672_hour  rides_previous_671_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
89300                      3.0                      0.0   
89301                      6.0                      4.0   
89302                      7.0                      2.0   
89303                      6.0                      5.0   
89304                      6.0                      6.0   

       rides_previous_670_hour  rides_previous_669_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
89300                      2.0                      3.0   
89301                      0.0                      0.0   
89302                      3.0                      4.0   
89303                      4.0                      3.0   
89304                      4.0                      2.0   

       rides_previous_668_hour  rides_previous_667_hour  \
0                          1.0                      1.0   
1                          0.0                      4.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
89300                      2.0                      3.0   
89301                      2.0                      0.0   
89302                      7.0                      4.0   
89303                      0.0                      3.0   
89304                      5.0                      1.0   

       rides_previous_666_hour  rides_previous_665_hour  \
0                          0.0                      2.0   
1                          1.0                      2.0   
2                          0.0                      2.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
89300                     13.0                      8.0   
89301                     14.0                      7.0   
89302                     10.0                      9.0   
89303                     11.0                     12.0   
89304                      0.0                      3.0   

       rides_previous_664_hour  rides_previous_663_hour  ...  \
0                          0.0                      0.0  ...   
1                          1.0                      2.0  ...   
2                          0.0                      0.0  ...   
3                          1.0                      1.0  ...   
4                          0.0                      0.0  ...   
...                        ...                      ...  ...   
89300                      9.0                      9.0  ...   
89301                      8.0                      4.0  ...   
89302                      7.0                     11.0  ...   
89303                      9.0                     10.0  ...   
89304                      5.0                      8.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        2.0                    0.0                    1.0   
1                        0.0                    0.0                    0.0   
2                        0.0                    1.0                    2.0   
3                        2.0                    1.0                    0.0   
4           

In [2]:
from datetime import datetime
from src.data_split import train_test_split

# Split data into train and test sets
X_train, y_train, X_test, y_test = train_test_split(
    df,
    cutoff_date=datetime(2022, 6, 1, 0, 0, 0),
    target_column_name='target_rides_next_hour'
)

print(f'{X_train.shape=}')
print(f'{y_train.shape=}')
print(f'{X_test.shape=}')
print(f'{y_test.shape=}')

X_train.shape=(32595, 674)
y_train.shape=(32595,)
X_test.shape=(56710, 674)
y_test.shape=(56710,)


In [3]:
def average_rides_last_4_weeks(X: pd.DataFrame) -> pd.DataFrame:
    """
    Adds one column with the average rides from
    - 7 days ago
    - 14 days ago
    - 21 days ago
    - 28 days ago

    Parameters
    ----------
    X : pd.DataFrame
        Input data

    Returns
    -------
    pd.DataFrame
        Input data with one additional column      
    """
    X['average_rides_last_4_weeks'] = 0.25*(
        X[f'rides_previous_{7*24}_hour'] + \
        X[f'rides_previous_{2*7*24}_hour'] + \
        X[f'rides_previous_{3*7*24}_hour'] + \
        X[f'rides_previous_{4*7*24}_hour']
    )
    return X

In [10]:
# transform the data
from sklearn.preprocessing import FunctionTransformer

add_feature_average_rides_last_4_weeks = FunctionTransformer(
    average_rides_last_4_weeks,validate=False)

In [11]:
add_feature_average_rides_last_4_weeks.fit_transform(X_train)

rides_previous_672_hour  rides_previous_671_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                     10.0                      7.0   
32591                      6.0                      5.0   
32592                      8.0                      6.0   
32593                      4.0                      2.0   
32594                      5.0                      7.0   

       rides_previous_670_hour  rides_previous_669_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                      4.0                      5.0   
32591                      5.0                      0.0   
32592                      2.0                      5.0   
32593                      5.0                      3.0   
32594                      4.0                      1.0   

       rides_previous_668_hour  rides_previous_667_hour  \
0                          1.0                      1.0   
1                          0.0                      4.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                     10.0                      4.0   
32591                      6.0                      4.0   
32592                      4.0                      3.0   
32593                      1.0                      7.0   
32594                      2.0                      5.0   

       rides_previous_666_hour  rides_previous_665_hour  \
0                          0.0                      2.0   
1                          1.0                      2.0   
2                          0.0                      2.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                      9.0                     15.0   
32591                      1.0                      3.0   
32592                      2.0                      3.0   
32593                     15.0                     18.0   
32594                     22.0                     20.0   

       rides_previous_664_hour  rides_previous_663_hour  ...  \
0                          0.0                      0.0  ...   
1                          1.0                      2.0  ...   
2                          0.0                      0.0  ...   
3                          1.0                      1.0  ...   
4                          0.0                      0.0  ...   
...                        ...                      ...  ...   
32590                     15.0                     13.0  ...   
32591                      3.0                     10.0  ...   
32592                     15.0                     14.0  ...   
32593                     17.0                     14.0  ...   
32594                     20.0                     24.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        2.0                    0.0                    1.0   
1                        0.0                    0.0                    0.0   
2                        0.0                    1.0                    2.0   
3                        2.0                    1.0                    0.0   
4           

In [12]:
from sklearn.base import BaseEstimator, TransformerMixin

class TemporalFeaturesEngineer(BaseEstimator, TransformerMixin):
    """
    Adds the following columns:
    - hour
    - day_of_week

    Parameters
    ----------
    BaseEstimator : [type]
        [description]
    TransformerMixin : [type]
        [description]

    Returns
    -------
    DataFrame with additional column
        Numeric column from datetime (day_of_week)
        Numeric column from datetime (hour) is DROPPED
    """

    def fit(self, X, y=None):
        return self
    
    def transform(self, X, y=None):
        X_ = X.copy()

        # generate numeric columns from datetime
        X_['hour'] = X_['pickup_hour'].dt.hour
        X_['day_of_week'] = X_['pickup_hour'].dt.dayofweek

        return X_.drop(columns=['pickup_hour'])

In [13]:
add_temporal_features = TemporalFeaturesEngineer()
add_temporal_features.fit_transform(X_train)

rides_previous_672_hour  rides_previous_671_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                     10.0                      7.0   
32591                      6.0                      5.0   
32592                      8.0                      6.0   
32593                      4.0                      2.0   
32594                      5.0                      7.0   

       rides_previous_670_hour  rides_previous_669_hour  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                      4.0                      5.0   
32591                      5.0                      0.0   
32592                      2.0                      5.0   
32593                      5.0                      3.0   
32594                      4.0                      1.0   

       rides_previous_668_hour  rides_previous_667_hour  \
0                          1.0                      1.0   
1                          0.0                      4.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                     10.0                      4.0   
32591                      6.0                      4.0   
32592                      4.0                      3.0   
32593                      1.0                      7.0   
32594                      2.0                      5.0   

       rides_previous_666_hour  rides_previous_665_hour  \
0                          0.0                      2.0   
1                          1.0                      2.0   
2                          0.0                      2.0   
3                          0.0                      0.0   
4                          0.0                      0.0   
...                        ...                      ...   
32590                      9.0                     15.0   
32591                      1.0                      3.0   
32592                      2.0                      3.0   
32593                     15.0                     18.0   
32594                     22.0                     20.0   

       rides_previous_664_hour  rides_previous_663_hour  ...  \
0                          0.0                      0.0  ...   
1                          1.0                      2.0  ...   
2                          0.0                      0.0  ...   
3                          1.0                      1.0  ...   
4                          0.0                      0.0  ...   
...                        ...                      ...  ...   
32590                     15.0                     13.0  ...   
32591                      3.0                     10.0  ...   
32592                     15.0                     14.0  ...   
32593                     17.0                     14.0  ...   
32594                     20.0                     24.0  ...   

       rides_previous_6_hour  rides_previous_5_hour  rides_previous_4_hour  \
0                        0.0                    1.0                    0.0   
1                        0.0                    0.0                    0.0   
2                        1.0                    2.0                    0.0   
3                        1.0                    0.0                    1.0   
4           

In [9]:
import lightgbm as lgb

In [14]:
from sklearn.pipeline import make_pipeline

pipeline = make_pipeline(
    add_feature_average_rides_last_4_weeks,
    add_temporal_features,
    lgb.LGBMRegressor()
)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('functiontransformer',
                 FunctionTransformer(func=<function average_rides_last_4_weeks at 0x138288310>)),
                ('temporalfeaturesengineer', TemporalFeaturesEngineer()),
                ('lgbmregressor', LGBMRegressor())])

In [15]:
predictions = pipeline.predict(X_test)

from sklearn.metrics import mean_absolute_error
test_mae = mean_absolute_error(y_test, predictions)
print(f'{test_mae=:.4f}')

test_mae=2.5947


### Feature Importance + Feature Engineering data - 4 weeks

In [33]:
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression

pipeline = make_pipeline(
    add_feature_average_rides_last_4_weeks,
    add_temporal_features,
    SelectKBest(f_regression, k=33), #f_regression is for regression, #k=33 is the num of features
    lgb.LGBMRegressor()
)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('functiontransformer',
                 FunctionTransformer(func=<function average_rides_last_4_weeks at 0x138288310>)),
                ('temporalfeaturesengineer', TemporalFeaturesEngineer()),
                ('selectkbest',
                 SelectKBest(k=33,
                             score_func=<function f_regression at 0x15f6739d0>)),
                ('lgbmregressor', LGBMRegressor())])

In [35]:
predictions = pipeline.predict(X_test)

from sklearn.metrics import mean_absolute_error
test_mae_importance = mean_absolute_error(y_test, predictions)
print(f'{test_mae_importance=:.4f}')

test_mae_importance=2.5695


### Summary 4 weeks

[w\o feature importance] 
- mae = 2.59
- feature_size = 676
- pipeline train time = 6.1 sec

[w feature importance] 
- test_mae_importance =2.5695 
- feature_size = 33
- pipeline train time = 2 sec

### 8 weeks of data

In [46]:
import pandas as pd
from src.paths import TRANSFORMED_DATA_DIR

df = pd.read_parquet(TRANSFORMED_DATA_DIR / 'tabular_data_2_months.parquet')
df

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       4.0                       6.0   
81881                       2.0                       6.0   
81882                       5.0                       1.0   
81883                       8.0                       6.0   
81884                       9.0                       4.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       0.0                       1.0   
81881                       4.0                       2.0   
81882                       3.0                       1.0   
81883                       3.0                       3.0   
81884                       4.0                       2.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       1.0                       7.0   
81881                       4.0                      10.0   
81882                       2.0                       8.0   
81883                       2.0                       8.0   
81884                       6.0                       3.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                      22.0                      18.0   
81881                      16.0                      17.0   
81882                      18.0                       9.0   
81883                      15.0                      16.0   
81884                       9.0                       4.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
81880                      25.0                      17.0  ...   
81881                      20.0                      12.0  ...   
81882                      18.0                      19.0  ...   
81883                       8.0                       8.0  ...   
81884                       5.0                       9.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        1.0                    2.0                    0.0   
1                        1.0                    0.0                    1.0   
2                        3.0                    

In [93]:
from datetime import datetime
from src.data_split import train_test_split

X_train, y_train, X_test, y_test = train_test_split(
    df,
    cutoff_date=datetime(2022, 6, 1, 0, 0, 0),
    target_column_name='target_rides_next_hour'
)

print(f'{X_train.shape=}')
print(f'{y_train.shape=}')
print(f'{X_test.shape=}')
print(f'{y_test.shape=}')

X_train.shape=(25175, 1346)
y_train.shape=(25175,)
X_test.shape=(56710, 1346)
y_test.shape=(56710,)


In [94]:
def average_rides_last_8_weeks(X: pd.DataFrame) -> pd.DataFrame:
    """
    Adds one column with the average rides from
    - 7 days ago
    - 14 days ago
    - 21 days ago
    - 28 days ago
    - 35 days ago
    - 42 days ago
    - 49 days ago
    - 56 days ago

    Parameters
    ----------
    X : pd.DataFrame
        Input data

    Returns
    -------
    pd.DataFrame
        Input data with one additional column      
    """
    X['average_rides_last_8_weeks'] = 0.125*(
        X[f'rides_previous_{7*24}_hour'] + \
        X[f'rides_previous_{2*7*24}_hour'] + \
        X[f'rides_previous_{3*7*24}_hour'] + \
        X[f'rides_previous_{4*7*24}_hour'] + \
        X[f'rides_previous_{5*7*24}_hour'] + \
        X[f'rides_previous_{6*7*24}_hour'] + \
        X[f'rides_previous_{7*7*24}_hour'] + \
        X[f'rides_previous_{8*7*24}_hour']
    )
    return X

In [95]:
# transform the data
from sklearn.preprocessing import FunctionTransformer

add_feature_average_rides_last_8_weeks = FunctionTransformer(
    average_rides_last_8_weeks,validate=False)

In [96]:
add_feature_average_rides_last_8_weeks.fit_transform(X_train)

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       9.0                       9.0   
25171                      11.0                       3.0   
25172                       6.0                       9.0   
25173                       6.0                       3.0   
25174                       8.0                       2.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       3.0                       4.0   
25171                       4.0                       3.0   
25172                       6.0                       2.0   
25173                       1.0                       0.0   
25174                       6.0                       4.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       6.0                       5.0   
25171                       3.0                       7.0   
25172                       8.0                       5.0   
25173                       5.0                       8.0   
25174                       3.0                       5.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                      23.0                      23.0   
25171                       2.0                       3.0   
25172                       6.0                       1.0   
25173                      20.0                      14.0   
25174                      16.0                      26.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
25170                      21.0                      16.0  ...   
25171                       6.0                       9.0  ...   
25172                       2.0                       7.0  ...   
25173                      24.0                      38.0  ...   
25174                      30.0                      31.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        1.0                    2.0                    0.0   
1                        1.0                    0.0                    1.0   
2                        3.0                    

In [97]:
from sklearn.base import BaseEstimator, TransformerMixin

class TemporalFeaturesEngineer(BaseEstimator, TransformerMixin):
    """
    Adds the following columns:
    - hour
    - day_of_week

    Parameters
    ----------
    BaseEstimator : [type]
        [description]
    TransformerMixin : [type]
        [description]

    Returns
    -------
    DataFrame with additional column
        Numeric column from datetime (day_of_week)
        Numeric column from datetime (hour) is DROPPED
    """

    def fit(self, X, y=None):
        return self
    
    def transform(self, X, y=None):
        X_ = X.copy()

        # generate numeric columns from datetime
        X_['hour'] = X_['pickup_hour'].dt.hour
        X_['day_of_week'] = X_['pickup_hour'].dt.dayofweek
        
        return X_.drop(columns=['pickup_hour'])

In [98]:
add_temporal_features = TemporalFeaturesEngineer()
add_temporal_features.fit_transform(X_train)

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       9.0                       9.0   
25171                      11.0                       3.0   
25172                       6.0                       9.0   
25173                       6.0                       3.0   
25174                       8.0                       2.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       3.0                       4.0   
25171                       4.0                       3.0   
25172                       6.0                       2.0   
25173                       1.0                       0.0   
25174                       6.0                       4.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                       6.0                       5.0   
25171                       3.0                       7.0   
25172                       8.0                       5.0   
25173                       5.0                       8.0   
25174                       3.0                       5.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
25170                      23.0                      23.0   
25171                       2.0                       3.0   
25172                       6.0                       1.0   
25173                      20.0                      14.0   
25174                      16.0                      26.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
25170                      21.0                      16.0  ...   
25171                       6.0                       9.0  ...   
25172                       2.0                       7.0  ...   
25173                      24.0                      38.0  ...   
25174                      30.0                      31.0  ...   

       rides_previous_6_hour  rides_previous_5_hour  rides_previous_4_hour  \
0                        2.0                    0.0                    0.0   
1                        0.0                    1.0                    1.0   
2                        2.0                    

In [99]:
from sklearn.pipeline import make_pipeline

pipeline = make_pipeline(
    add_feature_average_rides_last_8_weeks,
    add_temporal_features,
    lgb.LGBMRegressor()
)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('functiontransformer',
                 FunctionTransformer(func=<function average_rides_last_8_weeks at 0x162067a60>)),
                ('temporalfeaturesengineer', TemporalFeaturesEngineer()),
                ('lgbmregressor', LGBMRegressor())])

In [100]:
predictions = pipeline.predict(X_test)

from sklearn.metrics import mean_absolute_error
test_mae = mean_absolute_error(y_test, predictions)
print(f'{test_mae=:.4f}')

test_mae=2.7288


### Feature Importance + Feature Engineering data - 8 weeks

In [108]:
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression

pipeline = make_pipeline(
    add_feature_average_rides_last_8_weeks,
    add_temporal_features,
    SelectKBest(f_regression, k=40), #f_regression is for regression, #k=33 is the num of features
    lgb.LGBMRegressor()
)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('functiontransformer',
                 FunctionTransformer(func=<function average_rides_last_8_weeks at 0x162067a60>)),
                ('temporalfeaturesengineer', TemporalFeaturesEngineer()),
                ('selectkbest',
                 SelectKBest(k=40,
                             score_func=<function f_regression at 0x15f6739d0>)),
                ('lgbmregressor', LGBMRegressor())])

In [109]:
predictions = pipeline.predict(X_test)

from sklearn.metrics import mean_absolute_error
test_mae_importance = mean_absolute_error(y_test, predictions)
print(f'{test_mae_importance=:.4f}')

test_mae_importance=2.7380


### Summary 8 weeks
[w\o feature importance] 
- mae = 2.72
- feature_size = 1348
- pipeline train time = 8.5 sec

[w feature importance] 
- test_mae_importance = 2.73
- feature_size = 40
- pipeline train time = 3 sec

### Experiments section
To Do:
- [x] Experiment with 16 weeks (4 months) of data - to get the 70\30 split of the train and test data

In [110]:
import pandas as pd
from src.paths import TRANSFORMED_DATA_DIR

df = pd.read_parquet(TRANSFORMED_DATA_DIR / 'tabular_data_2_months.parquet')
df

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       4.0                       6.0   
81881                       2.0                       6.0   
81882                       5.0                       1.0   
81883                       8.0                       6.0   
81884                       9.0                       4.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       0.0                       1.0   
81881                       4.0                       2.0   
81882                       3.0                       1.0   
81883                       3.0                       3.0   
81884                       4.0                       2.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                       1.0                       7.0   
81881                       4.0                      10.0   
81882                       2.0                       8.0   
81883                       2.0                       8.0   
81884                       6.0                       3.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
81880                      22.0                      18.0   
81881                      16.0                      17.0   
81882                      18.0                       9.0   
81883                      15.0                      16.0   
81884                       9.0                       4.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
81880                      25.0                      17.0  ...   
81881                      20.0                      12.0  ...   
81882                      18.0                      19.0  ...   
81883                       8.0                       8.0  ...   
81884                       5.0                       9.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        1.0                    2.0                    0.0   
1                        1.0                    0.0                    1.0   
2                        3.0                    

In [111]:
from datetime import datetime
from src.data_split import train_test_split

X_train, y_train, X_test, y_test = train_test_split(
    df,
    cutoff_date=datetime(2022, 10, 1, 0, 0, 0),
    target_column_name='target_rides_next_hour'
)

print(f'{X_train.shape=}')
print(f'{y_train.shape=}')
print(f'{X_test.shape=}')
print(f'{y_test.shape=}')

X_train.shape=(57505, 1346)
y_train.shape=(57505,)
X_test.shape=(24380, 1346)
y_test.shape=(24380,)


In [112]:
def average_rides_last_8_weeks(X: pd.DataFrame) -> pd.DataFrame:
    """
    Adds one column with the average rides from
    - 7 days ago
    - 14 days ago
    - 21 days ago
    - 28 days ago
    - 35 days ago
    - 42 days ago
    - 49 days ago
    - 56 days ago

    Parameters
    ----------
    X : pd.DataFrame
        Input data

    Returns
    -------
    pd.DataFrame
        Input data with one additional column      
    """
    X['average_rides_last_8_weeks'] = 0.125*(
        X[f'rides_previous_{7*24}_hour'] + \
        X[f'rides_previous_{2*7*24}_hour'] + \
        X[f'rides_previous_{3*7*24}_hour'] + \
        X[f'rides_previous_{4*7*24}_hour'] + \
        X[f'rides_previous_{5*7*24}_hour'] + \
        X[f'rides_previous_{6*7*24}_hour'] + \
        X[f'rides_previous_{7*7*24}_hour'] + \
        X[f'rides_previous_{8*7*24}_hour']
    )
    return X

In [113]:
# transform the data
from sklearn.preprocessing import FunctionTransformer

add_feature_average_rides_last_8_weeks = FunctionTransformer(
    average_rides_last_8_weeks,validate=False)

In [114]:
add_feature_average_rides_last_8_weeks.fit_transform(X_train)

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       7.0                       5.0   
57501                       5.0                       1.0   
57502                       6.0                       3.0   
57503                      10.0                       2.0   
57504                      10.0                       7.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       1.0                       1.0   
57501                       1.0                       1.0   
57502                       3.0                       1.0   
57503                       1.0                       1.0   
57504                       6.0                       2.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       3.0                       4.0   
57501                       4.0                       7.0   
57502                       0.0                       3.0   
57503                       4.0                       3.0   
57504                       5.0                      11.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                      15.0                      17.0   
57501                       8.0                      17.0   
57502                       7.0                      20.0   
57503                      11.0                      13.0   
57504                       9.0                      14.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
57500                      21.0                       9.0  ...   
57501                      12.0                      15.0  ...   
57502                      12.0                      27.0  ...   
57503                       7.0                      15.0  ...   
57504                       8.0                       7.0  ...   

       rides_previous_7_hour  rides_previous_6_hour  rides_previous_5_hour  \
0                        1.0                    2.0                    0.0   
1                        1.0                    0.0                    1.0   
2                        3.0                    

In [115]:
from sklearn.base import BaseEstimator, TransformerMixin

class TemporalFeaturesEngineer(BaseEstimator, TransformerMixin):
    """
    Adds the following columns:
    - hour
    - day_of_week

    Parameters
    ----------
    BaseEstimator : [type]
        [description]
    TransformerMixin : [type]
        [description]

    Returns
    -------
    DataFrame with additional column
        Numeric column from datetime (day_of_week)
        Numeric column from datetime (hour) is DROPPED
    """

    def fit(self, X, y=None):
        return self
    
    def transform(self, X, y=None):
        X_ = X.copy()

        # generate numeric columns from datetime
        X_['hour'] = X_['pickup_hour'].dt.hour
        X_['day_of_week'] = X_['pickup_hour'].dt.dayofweek
        
        return X_.drop(columns=['pickup_hour'])

In [116]:
add_temporal_features = TemporalFeaturesEngineer()
add_temporal_features.fit_transform(X_train)

rides_previous_1344_hour  rides_previous_1343_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       7.0                       5.0   
57501                       5.0                       1.0   
57502                       6.0                       3.0   
57503                      10.0                       2.0   
57504                      10.0                       7.0   

       rides_previous_1342_hour  rides_previous_1341_hour  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       1.0                       1.0   
57501                       1.0                       1.0   
57502                       3.0                       1.0   
57503                       1.0                       1.0   
57504                       6.0                       2.0   

       rides_previous_1340_hour  rides_previous_1339_hour  \
0                           1.0                       1.0   
1                           0.0                       4.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                       3.0                       4.0   
57501                       4.0                       7.0   
57502                       0.0                       3.0   
57503                       4.0                       3.0   
57504                       5.0                      11.0   

       rides_previous_1338_hour  rides_previous_1337_hour  \
0                           0.0                       2.0   
1                           1.0                       2.0   
2                           0.0                       2.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
57500                      15.0                      17.0   
57501                       8.0                      17.0   
57502                       7.0                      20.0   
57503                      11.0                      13.0   
57504                       9.0                      14.0   

       rides_previous_1336_hour  rides_previous_1335_hour  ...  \
0                           0.0                       0.0  ...   
1                           1.0                       2.0  ...   
2                           0.0                       0.0  ...   
3                           1.0                       1.0  ...   
4                           0.0                       0.0  ...   
...                         ...                       ...  ...   
57500                      21.0                       9.0  ...   
57501                      12.0                      15.0  ...   
57502                      12.0                      27.0  ...   
57503                       7.0                      15.0  ...   
57504                       8.0                       7.0  ...   

       rides_previous_6_hour  rides_previous_5_hour  rides_previous_4_hour  \
0                        2.0                    0.0                    0.0   
1                        0.0                    1.0                    1.0   
2                        2.0                    

In [117]:
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression

pipeline = make_pipeline(
    add_feature_average_rides_last_8_weeks,
    add_temporal_features,
    SelectKBest(f_regression, k=40), #f_regression is for regression, #k=33 is the num of features
    lgb.LGBMRegressor()
)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('functiontransformer',
                 FunctionTransformer(func=<function average_rides_last_8_weeks at 0x16143e550>)),
                ('temporalfeaturesengineer', TemporalFeaturesEngineer()),
                ('selectkbest',
                 SelectKBest(k=40,
                             score_func=<function f_regression at 0x15f6739d0>)),
                ('lgbmregressor', LGBMRegressor())])

In [118]:
predictions = pipeline.predict(X_test)

from sklearn.metrics import mean_absolute_error
test_mae_importance = mean_absolute_error(y_test, predictions)
print(f'{test_mae_importance=:.4f}')

test_mae_importance=2.5951


### Summary 16 weeks (4 months) with 70\30 split
[w\o feature importance] 
- mae = xxx
- feature_size = xxx
- pipeline train time = xxx

[w feature importance] 
- test_mae_importance = 2.59
- feature_size = 40
- pipeline train time = 3 sec